lib/livebook_web/plugs/auth_plug.ex

```diff
@@ -16,36 +16,63 @@ defmodule LivebookWeb.AuthPlug do
   def init(opts), do: opts
 
   @impl true
-  def call(conn, _otps) do
-    case Application.get_env(:livebook, :token) do
-      nil -> conn
-      token -> token_authentication(conn, token)
+  def call(conn, opts) do
+    cond do
+      Application.fetch_env!(:livebook, :password_authentication) == true ->
+        password = Application.fetch_env!(:livebook, :password)
+        secret_authentication(conn, password, :password, opts)
+
+      Application.fetch_env!(:livebook, :token_authentication) == true ->
+        token = Application.fetch_env!(:livebook, :token)
+        secret_authentication(conn, token, :token, opts)
+
+      true ->
+        conn
     end
   end
 
-  defp token_authentication(conn, token) do
+  defp secret_authentication(conn, secret, type, opts) do
     # The user may run multiple Livebook instances on the same host
     # on different ports, so we scope the cookie name under port
-    token_cookie = "#{conn.port}:token"
-
-    conn = fetch_cookies(conn, signed: [token_cookie])
+    secret_cookie = "#{conn.port}:#{inspect(type)}"
 
-    param_token = Map.get(conn.query_params, "token")
-    cookie_token = conn.cookies[token_cookie]
+    conn = fetch_cookies(conn, signed: [secret_cookie])
+    param_secret = get_param_secret(conn, type)
+    cookie_secret = conn.cookies[secret_cookie]
 
     cond do
-      is_binary(param_token) and Plug.Crypto.secure_compare(param_token, token) ->
+      # check fresh secret
+      is_binary(param_secret) and Plug.Crypto.secure_compare(param_secret, secret) ->
         conn
-        |> put_resp_cookie(token_cookie, param_token, @cookie_opts)
+        |> put_resp_cookie(secret_cookie, param_secret, @cookie_opts)
         # Redirect to the same path without query params
-        |> redirect(to: conn.request_path)
+        |> redirect(to: opts[:path] || conn.request_path)
         |> halt()
 
-      is_binary(cookie_token) and Plug.Crypto.secure_compare(cookie_token, token) ->
+      # check secret stored in cookie
+      is_binary(cookie_secret) and Plug.Crypto.secure_compare(cookie_secret, secret) ->
+        conn
+
+      # on password type redirect to proper route
+      type == :password ->
         conn
+        |> redirect(to: "/authenticate")
+        |> halt()
```

Comment:
To stay in line with the token you can define `LivebookWeb.InvalidPasswordError` and then pattern match on that in `ErrorView`. This way we don't need the additional controller method just to render the form.

Reply:
Ah I guess the difference is that with this redirect we end up on the `/authenticate` path, so form submission naturally goes to the correct controller. But we may as well set the form target path explicitly.

Reply:
But the same can happen with token and, after all, the token submission can be invalid too. Maybe we should unify everything under `/auth`?

Reply:
One difference with token is that it preserves the URL you visited initially. This is not particularly relevant though, because the token error is gonna usually pop up when the user visits `localhost:8080` manually without the token.

Reply:
>  Maybe we should unify everything under `/auth`?

If we did that we would have to distinguish authentication mode both in the plug and in the controller. Not sure what's cleanest though.

Reply:
Discussing offline with @josevalim the idea would be to handle these somewhat separately:

* for token, the plug checks for the token and optionally sets it (so exactly how it was)
* for password, the plug only validates it and redirects to authentication (as you did), and then the controller handles the POST request on its own (setting the necessary cookies - instead of calling the plug)

On the plug side these flows are different, so having separate functions would make most sense :cat2: